File: lib/core/engine.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Core scanning engine for blackSQL
"""

import concurrent.futures
import time
import os
from urllib.parse import parse_qsl, urlparse

from ..utils.http_utils import RequestHandler
from ..utils.cli import print_status, progress_bar
from ..utils.validator import extract_params, parse_cookies, parse_post_data
from ..utils.logger import VulnerabilityLogger
from ..payloads.sql_payloads import ERROR_BASED, BOOLEAN_BASED, TIME_BASED, UNION_BASED, DB_FINGERPRINT, WAF_BYPASS
from ..techniques.error_based import ErrorBasedDetector
from ..techniques.boolean_based import BooleanBasedDetector
from ..techniques.time_based import TimeBasedDetector
from ..techniques.union_based import UnionBasedDetector
from ..techniques.extractor import DatabaseExtractor


class Scanner:
    """Main SQL injection scanning engine"""

    def __init__(self, url, params=None, data=None, cookies=None, threads=5,
                 timeout=10, proxy=None, level=1, dump=False, batch=False, logger=None):
        """
        Initialize the scanner

        Args:
            url (str): Target URL
            params (list, optional): List of parameters to scan
            data (str, optional): POST data string
            cookies (str, optional): HTTP cookie string
            threads (int): Number of threads to use
            timeout (float): Request timeout in seconds
            proxy (str, optional): Proxy URL
            level (int): Scan level (1-3)
            dump (bool): Whether to attempt database dumping
            batch (bool): Whether to use batch mode (no user input)
            logger (logging.Logger, optional): Logger instance
        """
        self.url = url
        self.params = params
        self.data_string = data
        self.cookie_string = cookies
        self.threads = threads
        self.timeout = timeout
        self.proxy = proxy
        self.level = level
        self.dump = dump
        self.batch = batch
        self.logger = logger

        # Parse data and cookies
        self.data = parse_post_data(data) if data else {}
        self.cookies = parse_cookies(cookies) if cookies else {}

        # Create request handler
        self.request_handler = RequestHandler(
            timeout=timeout,
            proxy=proxy,
            cookies=self.cookies
        )

        # Create vulnerability logger
        self.vuln_logger = VulnerabilityLogger()

        # Detect parameters if not provided
        if not self.params:
            self.params = list(extract_params(url).keys())

            # Add POST parameters if available
            if self.data:
                self.params.extend(list(self.data.keys()))

        # Prepare payloads based on scan level
        self.prepare_payloads()

        # Detected vulnerabilities
        self.vulnerabilities = []

    def prepare_payloads(self):
        """Prepare payloads based on scan level"""
        # Level 1: Basic payloads (about 25% of all payloads)
        # Level 2: Medium payloads (about 50% of all payloads)
        # Level 3: All payloads

        if self.level == 1:
            # Basic scan - fewer payloads
            self.error_payloads = ERROR_BASED[:7]
            self.boolean_payloads = BOOLEAN_BASED[:6]
            self.time_payloads = TIME_BASED[:4]
            self.union_payloads = []  # Skip UNION-based at level 1
        elif self.level == 2:
            # Medium scan
            self.error_payloads = ERROR_BASED[:15]
            self.boolean_payloads = BOOLEAN_BASED[:12]
            self.time_payloads = TIME_BASED[:10]
            # Skip UNION-based at level 2 (will run if others detect vulnerabilities)
            self.union_payloads = []
        else:
            # Level 3: Thorough scan - all payloads
            self.error_payloads = ERROR_BASED
            self.boolean_payloads = BOOLEAN_BASED
            self.time_payloads = TIME_BASED
            self.union_payloads = UNION_BASED  # Include UNION-based at level 3

    def scan_parameter(self, parameter):
        """
        Scan a single parameter for SQL injection

        Args:
            parameter (str): Parameter to scan

        Returns:
            dict: Scan results for this parameter
        """
        self.logger.info(f"Scanning parameter: {parameter}")

        # Determine if parameter is in URL or POST data
        is_post = parameter in self.data
        data = self.data if is_post else None

        # Results for this parameter
        results = {
            'parameter': parameter,
            'is_vulnerable': False,
            'techniques': [],
            'database_type': None
        }

        # 1. Check for Error-based SQL Injection
        self.logger.info(
            f"Testing Error-based SQL injection on parameter: {parameter}")
        error_detector = ErrorBasedDetector(
            self.url, self.request_handler, self.logger, self.vuln_logger)
        is_vuln_error, db_type_error, payload_error = error_detector.scan_parameter(
            parameter, self.error_payloads, is_post, data
        )

        if is_vuln_error:
            results['is_vulnerable'] = True
            results['techniques'].append('Error-based')
            results['database_type'] = db_type_error if db_type_error and db_type_error != 'general' else results['database_type']

        # 2. Check for Boolean-based SQL Injection
        self.logger.info(
            f"Testing Boolean-based SQL injection on parameter: {parameter}")
        boolean_detector = BooleanBasedDetector(
            self.url, self.request_handler, self.logger, self.vuln_logger)
        is_vuln_boolean, db_type_boolean, payload_boolean = boolean_detector.scan_parameter(
            parameter, self.boolean_payloads, is_post, data
        )

        if is_vuln_boolean:
            results['is_vulnerable'] = True
            results['techniques'].append('Boolean-based')
            results['database_type'] = db_type_boolean if db_type_boolean != 'Unknown' else results['database_type']

        # 3. Check for Time-based SQL Injection
        self.logger.info(
            f"Testing Time-based SQL injection on parameter: {parameter}")
        time_detector = TimeBasedDetector(
            self.url, self.request_handler, self.logger, self.vuln_logger)
        is_vuln_time, db_type_time, payload_time = time_detector.scan_parameter(
            parameter, self.time_payloads, is_post, data
        )

        if is_vuln_time:
            results['is_vulnerable'] = True
            results['techniques'].append('Time-based')
            results['database_type'] = db_type_time if db_type_time != 'Unknown' else results['database_type']

        # 4. Check for Union-based SQL Injection
        # Only run if other tests detected vulnerabilities or we're at level 3
        if results['is_vulnerable'] or self.level == 3:
            self.logger.info(
                f"Testing Union-based SQL injection on parameter: {parameter}")
            union_detector = UnionBasedDetector(
                self.url, self.request_handler, self.logger, self.vuln_logger)
            is_vuln_union, db_type_union, payload_union = union_detector.scan_parameter(
                parameter, is_post, data
            )

            if is_vuln_union:
                results['is_vulnerable'] = True
                results['techniques'].append('Union-based')
                results['database_type'] = db_type_union if db_type_union != 'Unknown' else results['database_type']

        # 5. Try to extract database information if parameter is vulnerable and dump is enabled
        if results['is_vulnerable'] and self.dump:
            self.logger.info(
                f"Attempting to extract database information from parameter: {parameter}")
            print_status(
                f"Attempting to extract database information from parameter: {parameter}", "info")

            extractor = DatabaseExtractor(
                self.url, parameter, results['database_type'],
                self.request_handler, self.logger, self.vuln_logger,
                is_post, data
            )

            extraction_results = extractor.extract_all()
            results['extraction'] = extraction_results

        # Store results
        if results['is_vulnerable']:
            self.vulnerabilities.append(results)

        return results

    def start(self):
        """Start scanning process"""
        self.logger.info(f"Starting scan with {self.threads} threads")
        print_status(f"Starting scan with {self.threads} threads", "info")

        start_time = time.time()

        # Check if we have parameters to scan
        if not self.params:
            self.logger.warning("No parameters found to scan")
            print_status(
                "No parameters found to scan. Please provide parameters with --params or use a URL with query parameters.", "warning")
            return

        # Display parameters to scan
        self.logger.info(f"Parameters to scan: {', '.join(self.params)}")
        print_status(f"Parameters to scan: {', '.join(self.params)}", "info")

        # Use ThreadPoolExecutor to scan parameters in parallel
        with concurrent.futures.ThreadPoolExecutor(max_workers=self.threads) as executor:
            # Submit tasks
            future_to_param = {executor.submit(
                self.scan_parameter, param): param for param in self.params}

            # Process results as they complete
            total_params = len(self.params)
            completed = 0

            for future in concurrent.futures.as_completed(future_to_param):
                param = future_to_param[future]
                try:
                    result = future.result()
                    completed += 1

                    # Update progress bar
                    progress_bar(completed, total_params,
                                 prefix=f'Progress:',
                                 suffix=f'Complete ({completed}/{total_params})',
                                 length=50)

                    # Report vulnerability if found
                    if result['is_vulnerable']:
                        print_status(
                            f"Parameter '{param}' is vulnerable to SQL injection", "vuln")
                        print_status(
                            f"Techniques: {', '.join(result['techniques'])}", "info")
                        print_status(
                            f"Database type: {result['database_type'] or 'Unknown'}", "info")
                except Exception as e:
                    self.logger.error(
                        f"Error scanning parameter {param}: {str(e)}")
                    print_status(
                        f"Error scanning parameter {param}: {str(e)}", "error")
                    completed += 1

                    # Update progress bar
                    progress_bar(completed, total_params,
                                 prefix=f'Progress:',
                                 suffix=f'Complete ({completed}/{total_params})',
                                 length=50)

        # Display results summary
        end_time = time.time()
        scan_duration = end_time - start_time

        print_status(f"Scan completed in {scan_duration:.2f} seconds", "info")
        print_status(f"Total parameters scanned: {len(self.params)}", "info")
        print_status(
            f"Vulnerable parameters found: {len(self.vulnerabilities)}", "info")

        # Export vulnerability results if found
        if self.vulnerabilities:
            # Create output directory if it doesn't exist
            os.makedirs("output", exist_ok=True)

            # Generate timestamp-based filename
            timestamp = time.strftime("%Y%m%d_%H%M%S")
            json_file = f"output/blacksql_results_{timestamp}.json"
            csv_file = f"output/blacksql_results_{timestamp}.csv"

            # Export results
            self.vuln_logger.export_to_json(json_file)
            self.vuln_logger.export_to_csv(csv_file)

            print_status(f"Results exported to JSON: {json_file}", "success")
            print_status(f"Results exported to CSV: {csv_file}", "success")

        return self.vulnerabilities

File: src/engine.rs
//! Scanner engine - port of legacy/lib/core/engine.py

use std::collections::HashMap;
use std::sync::mpsc;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Instant;

use crate::cli::{print_status, progress_bar, Status};
use crate::http_client::RequestHandler;
use crate::payloads::{prepare_payloads, PreparedPayloads};
use crate::techniques::{scan_boolean_based, scan_error_based, scan_time_based, scan_union_based, extract_all, ScanContext};
use crate::waf_detector;

/// Result for one parameter (same shape as Python engine result).
#[derive(Clone, Debug)]
pub struct ParamResult {
    pub parameter: String,
    pub is_vulnerable: bool,
    pub techniques: Vec<String>,
    pub database_type: Option<String>,
    /// Set when --dump and vulnerable (Phase 6).
    pub extraction: Option<crate::techniques::ExtractionResults>,
}

/// Scanner config (same as Python Scanner.__init__ args).
pub struct ScannerConfig {
    pub url: String,
    pub params: Vec<String>,
    pub data: HashMap<String, String>,
    pub cookies: HashMap<String, String>,
    pub threads: u32,
    pub timeout: f64,
    pub proxy: Option<String>,
    pub level: u8,
    pub dump: bool,
    pub batch: bool,
}

/// Run all techniques on one parameter. Order: error → boolean → time; if vulnerable or level 3 → union.
fn scan_one_parameter(
    url: &str,
    handler: &RequestHandler,
    parameter: &str,
    data: &HashMap<String, String>,
    level: u8,
    dump: bool,
    prepared: &PreparedPayloads,
) -> ParamResult {
    let is_post = data.contains_key(parameter);
    let data_opt = if is_post { Some(data) } else { None };
    let ctx = ScanContext {
        url,
        handler,
        is_post,
        data: data_opt,
    };

    let mut techniques = Vec::new();
    let mut database_type: Option<String> = None;

    let r_error = scan_error_based(&ctx, parameter, &prepared.error);
    if r_error.vulnerable {
        techniques.push("Error-based".to_string());
        if r_error.db_type.as_deref() != Some("Unknown") {
            database_type = r_error.db_type.clone();
        }
    }

    let r_bool = scan_boolean_based(&ctx, parameter, &prepared.boolean);
    if r_bool.vulnerable {
        techniques.push("Boolean-based".to_string());
        if r_bool.db_type.as_deref() != Some("Unknown") {
            database_type = database_type.or(r_bool.db_type.clone());
        }
    }

    let r_time = scan_time_based(&ctx, parameter, &prepared.time);
    if r_time.vulnerable {
        techniques.push("Time-based".to_string());
        if r_time.db_type.as_deref() != Some("Unknown") {
            database_type = database_type.or(r_time.db_type.clone());
        }
    }

    let run_union = !techniques.is_empty() || level == 3;
    if run_union {
        let r_union = scan_union_based(&ctx, parameter);
        if r_union.vulnerable {
            techniques.push("Union-based".to_string());
            if r_union.db_type.as_deref() != Some("Unknown") {
                database_type = database_type.or(r_union.db_type.clone());
            }
        }
    }

    let is_vulnerable = !techniques.is_empty();
    let extraction = if is_vulnerable && dump {
        print_status(
            &format!("Attempting to extract database information from parameter: {}", parameter),
            Status::Info,
        );
        let ctx = ScanContext { url, handler, is_post, data: data_opt };
        Some(extract_all(&ctx, parameter, database_type.as_deref()))
    } else {
        None
    };

    ParamResult {
        parameter: parameter.to_string(),
        is_vulnerable,
        techniques,
        database_type,
        extraction,
    }
}

/// Start the scan: WAF check, parallel workers, progress bar, summary.
pub fn run_scan(config: ScannerConfig) -> Vec<ParamResult> {
    let level = config.level.clamp(1, 3);
    let mut use_waf_bypass = true;

    let handler = match RequestHandler::with_cookies(
        config.timeout,
        config.proxy.as_deref(),
        &config.cookies,
    ) {
        Ok(h) => Arc::new(h),
        Err(e) => {
            print_status(&format!("Failed to create HTTP client: {}", e), Status::Error);
            return Vec::new();
        }
    };

    if config.params.is_empty() {
        print_status(
            "No parameters to scan. Use --params or a URL with query parameters or --data.",
            Status::Warning,
        );
        return Vec::new();
    }

    print_status("Checking if target is protected by a WAF...", Status::Info);
    let (is_waf, _) = waf_detector::check_target(handler.as_ref(), &config.url);
    if is_waf {
        use_waf_bypass = true;
    }

    let mut prepared = prepare_payloads(level, use_waf_bypass);
    if is_waf {
        prepared = prepare_payloads(level, true);
    }

    print_status(&format!("Parameters to scan: {}", config.params.join(", ")), Status::Info);
    print_status(&format!("Starting scan with {} threads", config.threads), Status::Info);

    let start = Instant::now();
    let n = config.params.len();
    let num_workers = config.threads as usize;
    let num_workers = num_workers.min(n).max(1);

    let queue: Arc<Mutex<Vec<String>>> = Arc::new(Mutex::new(config.params.clone()));
    let (tx, rx) = mpsc::channel::<ParamResult>();

    let url = config.url.clone();
    let data = config.data.clone();

    let dump = config.dump;
    for _ in 0..num_workers {
        let url_c = url.clone();
        let handler_c = Arc::clone(&handler);
        let data_c = data.clone();
        let prepared_c = prepared.clone();
        let queue_c = Arc::clone(&queue);
        let tx_c = tx.clone();
        thread::spawn(move || {
            loop {
                let param = {
                    let mut q = queue_c.lock().unwrap();
                    q.pop()
                };
                let param = match param {
                    Some(p) => p,
                    None => break,
                };
                let result = scan_one_parameter(
                    &url_c,
                    handler_c.as_ref(),
                    &param,
                    &data_c,
                    level,
                    dump,
                    &prepared_c,
                );
                let _ = tx_c.send(result);
            }
        });
    }
    drop(tx);

    let mut vulnerabilities = Vec::new();
    for (i, received) in rx.iter().enumerate() {
        let done = i + 1;
        progress_bar(
            done,
            n,
            "Progress:",
            &format!("Complete ({}/{})", done, n),
        );
        if received.is_vulnerable {
            vulnerabilities.push(received);
        }
    }

    let elapsed = start.elapsed().as_secs_f64();
    print_status(&format!("Scan completed in {:.2} seconds", elapsed), Status::Info);
    print_status(&format!("Total parameters scanned: {}", n), Status::Info);
    print_status(
        &format!("Vulnerable parameters found: {}", vulnerabilities.len()),
        Status::Info,
    );

    for v in &vulnerabilities {
        print_status(
            &format!("  {} [{}] db={:?}", v.parameter, v.techniques.join(", "), v.database_type),
            Status::Vuln,
        );
    }

    if !vulnerabilities.is_empty() {
        if let Err(e) = crate::output::export_results(&config.url, &vulnerabilities) {
            print_status(&format!("Failed to export results: {}", e), Status::Error);
        }
    }

    vulnerabilities
}
